### ISSUE TITLE
关于models.py 几个问题

### ISSUE THREAD

**Issue by SvenNJ** [kwai/DouZero#50]
`
self.lstm = nn.LSTM(162, 128, batch_first=True)

self.dense1 = nn.Linear(373 + 128, 512) # 输入501，输出 512
self.dense2 = nn.Linear(512, 512)
self.dense3 = nn.Linear(512, 512)
self.dense4 = nn.Linear(512, 512)
self.dense5 = nn.Linear(512, 512)
self.dense6 = nn.Linear(512, 1)
`

1、这个输入维度是162 ，我理解的是 z_batch 有 num_legal_actions行-5列-162个元素，也就是对应输入的162，不知道是不是这个意思
2、隐藏维度128 不清楚 有什么含义，可以随意设置吗
3、这里设计了6层全连接网路，不清楚 有什么含义，可以随意设置吗

**Comment by daochenzha** [comment]
@SvenNJ 162是纬度。128是hidden dimension，可以换。6层也可以调，用更多层也可以

**Comment by SvenNJ** [comment]
> @SvenNJ 162是纬度。128是hidden dimension，可以换。6层也可以调，用更多层也可以

就是不清楚这里设置 128 隐藏维度的依据是什么， 还有设置6层全连接网络 的依据。

**Comment by daochenzha** [comment]
@SvenNJ 128是网络参数数量 一般越大网络容量越大 但是训练起来更慢

**Comment by SvenNJ** [comment]
好的，非常感谢

**Comment by SvenNJ** [comment]
`
lstm_out, (h_n, _) = self.lstm(z)
lstm_out = lstm_out[:,-1,:]
x = torch.cat([lstm_out,x], dim=-1)
x = self.dense1(x)
x = torch.relu(x)
`

还有 2个问题：
1、这个参数z是 最近15手 出牌记录，三维张量tensor，最终代表的 合法牌型 行(一维)， 5列(也是3个人5轮历史出牌 二维)，162(三维， 162是3个人的历史出牌，54 x 3)

目前我看代码是理解到这里，问题是不清楚为什么使用这个z变量 来初始化 `lstm_out, (h_n, _) = self.lstm(z)`， 感觉是模拟人类打牌，根据历史出牌来作为 本次出牌的 【输入】， 具体 输入 数据 怎么来结合 下面的 牌 这里搞不懂

2、这里的参数x， 也就是x_batch ，依次是 my_handcards_batch、other_handcards_batch、last_action_batch、landlord_up_played_cards_batch、landlord_down_played_cards_batch、landlord_up_num_cards_left_batch、landlord_down_num_cards_left_batch、bomb_num_batch、my_action_batch， 二维张量tensor，代表的 合法牌型 行(一维)，373列(二维，前面这些数据拼接)， 问题是 搞不清楚 这些数据 具体起到了什么作用，或者说顺序上有什么要求，还有 如果加一些其他的类型牌 或者减少一些类型的数据 ，有什么影响？

问题1的历史出牌  和 问题2的数据 结合 torch.cat 起来，给到nn.Linear 和 激活torch.relu，具体的逻辑思维是怎么样的，有什么样的作用 和 理论依据。

我发现这个models.py 是这个项目的 最核心 和关键的部分，目前我只能看懂是怎么运行的，却不明白 为什么这么做，依据和理论基础我这里比较缺少。

`
网络的输入是状态和动作，输出是 公式 值。动作就是简单地用上面的方式进行编码。状态包括两部分：一部分是当前能看到的信息，包括手牌、其他玩家出的牌、上家的牌等特征矩阵以及其他玩家手牌数量和炸弹数量的0/1编码；另一部分是历史出牌信息，我们用LSTM网络进行编码。最后特征经过6层全连接网路得到 公式 值
`

包括看到这里的介绍，能明其形 但不明其意图，如果自己来做 其他类型 卡牌 AI学习 就可能会出问题。 望不吝赐教。

**Comment by daochenzha** [comment]
@SvenNJ 

1. 这里不是初始化，是把feature输入到lstm
2. 它们原来是矩阵，但是因为我们用的MLP，所以顺序没有影响

nn.Linear 加上relu激活函数是常规操作，具体可以看神经网络相关资料

**Comment by SvenNJ** [comment]
好的，我研究下
